oauth2_http/java/com/google/auth/oauth2/ImpersonatedCredentials.java

```diff
@@ -213,6 +291,76 @@ int getLifetime() {
         ImmutableMap.of("delegates", this.delegates));
   }
 
+  /**
+   * Returns impersonation account credentials defined by JSON using the format generated by gCloud.
+   *
+   * @param json a map from the JSON representing the credentials
+   * @param transportFactory HTTP transport factory, creates the transport used to get access tokens
+   * @return the credentials defined by the JSON
+   * @throws IOException if the credential cannot be created from the JSON.
+   */
+  static ImpersonatedCredentials fromJson(
+      Map<String, Object> json, HttpTransportFactory transportFactory) throws IOException {
+
+    checkNotNull(json);
+    checkNotNull(transportFactory);
+
+    String serviceAccountImpersonationUrl = (String) json.get("service_account_impersonation_url");
+    String targetPrincipal = extractTargetPrincipal(serviceAccountImpersonationUrl);
+
+    List<String> delegates = (List<String>) json.get("delegates");
```

Comment:
It looks like they're a lot of potential ClassCastExceptions here that are not being handled. Where does the json object come from?

Reply:
The json object is passed from https://github.com/googleapis/google-auth-library-java/blob/a5150b5b677cbb7e5f03e3ba9b43c5d7b030015e/oauth2_http/java/com/google/auth/oauth2/GoogleCredentials.java#L149 and it reads credential json file from disk. The json file is generated by gcloud which should be well-formatted and user shouldn't manually edit the file. We have done something similar in https://github.com/googleapis/google-auth-library-java/blob/a5150b5b677cbb7e5f03e3ba9b43c5d7b030015e/oauth2_http/java/com/google/auth/oauth2/ExternalAccountCredentials.java#L269

Reply:
The code should check this input and be prepared to handle a malformed or invalid file. It does not know where the file came from or how it might have been modified since it was generated. Do not assume it has the format you expect. Even if it's right today, the output of gcloud could change in the future.

If other code already makes this mistake, it should be fixed too.

Reply:
fair enough, thanks for pointing this out! I added the exception handling and added a test to cover it. Please take another look.